### PR TITLE
Deprecation Staging

### PR DESCRIPTION
**Pull request by mehulkar**
Thanks for @pzuraq,@rwjblue, @wycats, and the rest of the strike team for fleshing out these ideas and letting me play scribe!

[rendered](https://github.com/mehulkar/rfcs/blob/mk/deprecation-staging/text/0649-deprecation-staging.md)

### CODE REVIEW

**Comment by rwjblue** [comment]
Thanks for working on this @mehulkar!

**Comment by rwjblue** [comment]
Just FYI we discussed this at today's core team meeting. Nothing really to report (I don't think most folks have had a chance to review yet), but folks were excited about the progress. 😸

**Comment by camerondubas** [comment]
Nice! I don't have much to add here, although I think this is a nice progression of the deprecation system. 

Could the [rendered link](https://github.com/mehulkar/rfcs/blob/mk/deprecation-staging/text/0000-deprecation-staging.md) be added to the PR description? Thanks!

**Comment by rwjblue** [comment]
@camerondubas - Yep, thanks for pointing that out!

File: text/0000-deprecation-staging.md [hunk 1]
```diff
@@ -0,0 +1,652 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
```
Comment (rwjblue) [inline text/0000-deprecation-staging.md:1]: Can you rename the file to `text/0649-deprecation-staging.md`?

Comment (mehulkar) [reply text/0000-deprecation-staging.md]: done and done

File: text/0649-deprecation-staging.md [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+
+- by omitting a declaration
+- by passing an object conforming to the `AddonDeprecationConfig` interface.
+
+If a declaration is omitted, Ember will assume that the app is compliant with
+all deprecations in the "required" stage, but will not opt the app into any assertions.
+This ensures backwards compatibility with apps that do not add this config.
+
+If a `AddonDeprecationConfig` object is passed, Ember will assume that the app
+is compliant with deprecations of the configured `stage` and later, and the
```
Comment (mehulkar) [inline text/0649-deprecation-staging.md:262]:
```suggestion
is compliant with deprecations of the configured `stage`, and the
```

File: text/0649-deprecation-staging.md [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+    }
+  }
+}
+```
+
+### Configuring Addons
+
+Addons are also able to declare their compliance with other libraries in the same
+way as above. A declaration from an addon tells Ember that its _parent_ addon or application
+cannot be declare compliance to a higher version.
```
Comment (mehulkar) [inline text/0649-deprecation-staging.md:405]:
```suggestion
cannot declare compliance to a higher version.
```

File: text/0000-deprecation-staging.md (hunk 1 above)
Comment (rwjblue) [reply text/0000-deprecation-staging.md]: Thanks!

**Comment by rwjblue** [comment]
We chatted about this today in the core team meeting. The team is very happy with the overall design and we are eager to move forward. The one area that was still a bit of a concern was around the stage names. We are pretty happy with `available` but there was some back and forth about the `required` name. After we went back and forth a bit, we'd like to change `required` to `enabled`. Other than that one change, I think we are collectively very 👍 and would like to move forward!

**Comment by mehulkar** [comment]
🥳 great to hear! Enabled seems good to me too, I've updated the text.

**Comment by rwjblue** [comment]
Thanks for updating that @mehulkar! We discussed this in the most recent core team meeting, and have decided to move this into final comment period.

**Comment by rwjblue** [comment]
Chatted about this in the core team meeting today, we are still super happy with the direction forward and are ready to land this one!

**Pull request merged**